test/conformance_test.go

```diff
@@ -38,33 +42,119 @@ func TestTaskRun(t *testing.T) {
+		}},
+	}, {
+		name:   "failed-task-run",
+		trName: "failed-echo-hello-task-run",
+		tr: &v1beta1.TaskRun{
+			ObjectMeta: metav1.ObjectMeta{Name: "failed-echo-hello-task-run", Namespace: namespace},
+			Spec: v1beta1.TaskRunSpec{
+				TaskSpec: &v1beta1.TaskSpec{
+					Steps: []v1beta1.Step{{Container: corev1.Container{
+						Image:   "busybox",
+						Command: []string{"/bin/sh"},
+						Args:    []string{"-c", "echo hello"},
+					}}, {Container: corev1.Container{
+						Image:   "busybox",
+						Command: []string{"/bin/sh"},
+						Args:    []string{"-c", "exit 1"},
+					}}, {Container: corev1.Container{
+						Image:   "busybox",
+						Command: []string{"/bin/sh"},
+						Args:    []string{"-c", "sleep 30s"},
+					}}},
+				},
+			},
+		},
+		fn:                      TaskRunFailed,
+		expectedConditionStatus: corev1.ConditionFalse,
+		expectedStepState: []v1beta1.StepState{{
+			ContainerState: corev1.ContainerState{
+				Terminated: &corev1.ContainerStateTerminated{
+					ExitCode: 0,
+					Reason:   "Completed",
+				},
+			},
+			Name:          "unnamed-0",
+			ContainerName: "step-unnamed-0",
+		}, {
+			ContainerState: corev1.ContainerState{
+				Terminated: &corev1.ContainerStateTerminated{
+					ExitCode: 1,
+					Reason:   "Error",
+				},
+			},
+			Name:          "unnamed-1",
+			ContainerName: "step-unnamed-1",
+		}, {
+			ContainerState: corev1.ContainerState{
+				Terminated: &corev1.ContainerStateTerminated{
+					ExitCode: 1,
+					Reason:   "Error",
+				},
+			},
+			Name:          "unnamed-2",
+			ContainerName: "step-unnamed-2",
+		}},
+	}} {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Logf("Creating TaskRun %s", tc.trName)
+			if _, err := c.TaskRunClient.Create(ctx, tc.tr, metav1.CreateOptions{}); err != nil {
+				t.Fatalf("Failed to create TaskRun `%s`: %s", tc.trName, err)
+			}
 
-	t.Logf("Creating TaskRun %s", trName)
-	if _, err := c.TaskRunClient.Create(ctx, tr, metav1.CreateOptions{}); err != nil {
-		t.Fatalf("Failed to create TaskRun `%s`: %s", trName, err)
-	}
+			if err := WaitForTaskRunState(ctx, c, tc.trName, tc.fn(tc.trName), "WaitTaskRunDone"); err != nil {
+				t.Errorf("Error waiting for TaskRun to finish: %s", err)
+				return
+			}
+			tr, err := c.TaskRunClient.Get(ctx, tc.trName, metav1.GetOptions{})
+			if err != nil {
+				t.Fatalf("Failed to get TaskRun `%s`: %s", tc.trName, err)
+			}
+			checkRequiredFieldsInTaskRun(t, tr, tc.expectedConditionStatus)
 
-	if err := WaitForTaskRunState(ctx, c, trName, TaskRunSucceed(trName), "WaitTaskRunDone"); err != nil {
-		t.Errorf("Error waiting for TaskRun to finish: %s", err)
-		return
-	}
-	tr, err := c.TaskRunClient.Get(ctx, trName, metav1.GetOptions{})
-	if err != nil {
-		t.Fatalf("Failed to get TaskRun `%s`: %s", trName, err)
+			ignoreTerminatedFields := cmpopts.IgnoreFields(corev1.ContainerStateTerminated{}, "StartedAt", "FinishedAt", "ContainerID")
+			ignoreStepFields := cmpopts.IgnoreFields(v1beta1.StepState{}, "ImageID")
```

Comment:
ImageID _is_ REQUIRED by the spec. It's fine to ignore it for this change, but we should have a TODO to fix it before considering these tests done.

To make it easier to test, we can specify each step image by digest (`busybox@sha256:a9286defaba7b3a519d585ba0e37d0b2cbee74ebfe590960b0b1d6a5e97d1e1d`) and the step's imageID should be the digest portion (`sha256:a9286defaba7b3a519d585ba0e37d0b2cbee74ebfe590960b0b1d6a5e97d1e1d`)

That can happen in another change, but we should have a TODO for now.

Reply:
Sure. I added the check for ImageID in this change.